pkg/workflows/upgrade.go

```diff
@@ -119,37 +120,50 @@ func (s *setupAndValidateTasks) Run(ctx context.Context, commandContext *task.Co
 		commandContext.SetError(err)
 		return nil
 	}
+	currentSpec, err := commandContext.ClusterManager.GetCurrentClusterSpec(ctx, commandContext.ManagementCluster, commandContext.ClusterSpec.Cluster.Name)
+	if err != nil {
+		commandContext.SetError(err)
+		return &CollectDiagnosticsTask{}
+	}
+	commandContext.CurrentClusterSpec = currentSpec
+	s.FirstRunComplete = true
 
 	return &updateSecrets{}
 }
 
 func (s *setupAndValidateTasks) validations(ctx context.Context, commandContext *task.CommandContext) []validations.Validation {
-	return []validations.Validation{
+	result := []validations.Validation{
 		func() *validations.ValidationResult {
 			return &validations.ValidationResult{
 				Name: fmt.Sprintf("%s Provider setup is valid", commandContext.Provider.Name()),
 				Err:  commandContext.Provider.SetupAndValidateUpgradeCluster(ctx, commandContext.ManagementCluster, commandContext.ClusterSpec),
 			}
 		},
-		func() *validations.ValidationResult {
+	}
+	if !s.FirstRunComplete {
+		result = append(result, func() *validations.ValidationResult {
 			return &validations.ValidationResult{
 				Name: "upgrade preflight validations pass",
 				Err:  commandContext.Validations.PreflightValidations(ctx),
 			}
-		},
+		})
 	}
+	return result
 }
 
 func (s *setupAndValidateTasks) Name() string {
 	return "setup-and-validate"
 }
 
 func (s *setupAndValidateTasks) Restore(ctx context.Context, commandContext *task.CommandContext, completedTask *task.CompletedTask) (task.Task, error) {
-	return &updateSecrets{}, nil
+	s.FirstRunComplete = true
+	return s.Run(ctx, commandContext), nil
 }
 
 func (s *setupAndValidateTasks) Checkpoint() *task.CompletedTask {
-	return nil
+	return &task.CompletedTask{
```

Comment:
I took a peek at the `CompletedTask` and what not and it feels a little redundant because its a struct with a single field of type `interface{}`. Can we not use `interface{}`?

Reply:
We could. However, having it as a struct will allow us to build on top of it later if we want to retain even more information about the completed task